### ISSUE TITLE
Why "old" digest functions are deprecated?

### ISSUE THREAD

**Issue by tawmoto** [openssl/openssl#12260]
Hi, I am preparing a slow switch from openssl 1.1 to 3.0 and I know that the "old" digest functions are not recommended and deprecated in 3.0.
I am talking about this for example: https://www.openssl.org/docs/manmaster/man3/MD5_Init.html

I will switch to the new `EVP` functions, of course, but I was wondering why are the old ones deprecated? Because I understood that internally they can still be used...
Because for most of us, the old ones were easier to use `::MD5(...)` instead of calling 5 functions: `EVP` functions + `MD_CTX_new/free`

Other than flexibility, maybe a security reason or smth?

Thank you

**Comment by mattcaswell** [comment]
For a number of reasons. For example:

1) In some cases you get sub-optimal implementations with the low-level APIs vs the EVP APIs. An example from the "cipher" world (but the same concept applies to digests) is the low-level function AES_encrypt. If you call that you will never get the AESNI optimized version which may be available on your platform.

2) We want to encourage applications to use a consistent API for all types of digests/ciphers etc. This makes it much easier for everyone to update code as security advice changes. For example you mention the MD5 digest APIs. MD5 is no longer recommended. It's a lot harder to update your code to use some other digest if you've used the low-level APIs vs the high level ones. Looking into the future imagine some algorithm has some major flaw discovered in it that requires a rapid migration away from it to some other algorithm. We want the OpenSSL ecosystem to be as agile as possible to be able to deal with that.

3) The old APIs no longer fits architecturally with how OpenSSL 3.0 works. All algorithm implementations are now made available by "providers" in OpenSSL 3.0 - for example we have the "default" provider, the "fips" provider and a "legacy" provider.  The low level APIs circumvent all of that which means we have to maintain 2 ways of doing everything (3 actually when you bring ENGINEs into the picture as well). This leads to unnecessary code bloat and complexity....which is definitely something you want to avoid in a crypto library. Ideally we would have removed the old ways completely - but that would have been too big a breaking change.

**Comment by tawmoto** [comment]
Thank you for the explanations @mattcaswell.

**Comment by mspncp** [comment]
@mattcaswell what a [nice reply](https://github.com/openssl/openssl/issues/12260#issuecomment-649404727)! I think it deserves a place in the wiki.

**Comment by richsalz** [comment]
Yes, it's a great reply.  Also put it in the FAQ?

**Comment by romen** [comment]
I believe it is also a good idea to point at a few resources showing how insecure MD5 and others have been long deprecated although still in widespread use, in part because of a lack of agility in existing applications due to scarce agility in popular legacy crypto APIs (including OpenSSL low level ones).

- <https://shattered.io/> makes a good job at making easy to understand how broken SHA-1 (and even more MD5) is (SHAttered is from 2017, but SHA-1 attacks have been a reality since 2005, and NIST officially deprecated it already in 2011)
- <https://en.wikipedia.org/wiki/MD5> has also a very good summary. Basically broken since 1996, and repeatedly abused by white-hat and black-hat attackers violating both collision and preimage reistance.

**Comment by tawmoto** [comment]
@romen I know, MD5 was just an example that came to my mind. :-)

**Comment by romen** [comment]
> @romen I know, MD5 was just the first example that came to my mind. :-), I am using SHA256.

Yes, my comment was directed more towards enriching the very good answer by @mattcaswell in the optic of making it a part of a Wiki or FAQ entry (and for those that reach this answer searching on github issues).

I think we have a responsibility to sensitize our users regarding the level of security risks of using the legacy API and the legacy implementations: as a security researcher I would go even as far as saying that, looking at current —after _decades_ of deprecation!— adoption rates in the wild of MD5 and SHA-1 (or DES to venture into other domains), I feel —_personally and not as a voice for the OpenSSL project_— this is an area where I would prefer to ditch legacy to force existing applications to migrate to newer and more secure choices.

The OpenSSL project can't yet remove these APIs due to the promises it made, so I believe all members of the project share the burden of raising awareness of the security risks we are exposing the users to when they pick legacy APIs.
Even more given the fact that, as long as the compiler or the linker does not throw a warning —given how conservative we as a project are even with applying deprecations!—, maintainers of existing applications could be even unaware that their applications is insecure, and users of such applications can be even less aware of it.

**Comment by t8m** [comment]
>     * https://en.wikipedia.org/wiki/MD5 has also a very good summary. Basically broken since 1996, and repeatedly abused by white-hat and black-hat attackers violating both collision and preimage reistance.

Violating preimage resistance? Really? Can you point to any practical preimage attack on MD5?
Wikipedia talks only about: "In April 2009, an attack against MD5 was published that breaks MD5's preimage resistance. This attack is only theoretical, with a computational complexity of 2^123.4 for full preimage." That does not really look practical.

You probably mean chosen prefix collision attacks which are much more useful than the generic collision attacks. And yes both SHA-1 and MD5 are susceptible to this attack and moreover their Merkle-Damgard construction also provides ability to choose the suffix.

**Comment by romen** [comment]
> > ```
> > * https://en.wikipedia.org/wiki/MD5 has also a very good summary. Basically broken since 1996, and repeatedly abused by white-hat and black-hat attackers violating both collision and preimage reistance.
> > ```
> 
> Violating preimage resistance? Really? Can you point to any practical preimage attack on MD5?
> Wikipedia talks only about: "In April 2009, an attack against MD5 was published that breaks MD5's preimage resistance. This attack is only theoretical, with a computational complexity of 2^123.4 for full preimage." That does not really look practical.
> 
> You probably mean chosen prefix collision attacks which are much more useful than the generic collision attacks. And yes both SHA-1 and MD5 are susceptible to this attack and moreover their Merkle-Damgard construction also provides ability to choose the suffix.

@t8m I did not intend to claim anything more than it is written on that Wikipedia entry, and I did not distinguish between more or less practical attacks.
Anyway, the same authors of that paper, [published further improvements in 2013](http://ieeexplore.ieee.org/document/7223160/), but I am not aware of further published research on breaking its preimage resistance.

I am not saying the attack is now practical, but then again the fact that no further published results seem to be available does not mean that there aren't more viable practical attacks.

One has also to consider that there might be little incentive for public research to venture further into the topic, as it could be seen as a waste of resources: the prospective impact of research results attacking something that the research community considers a relic of an age past and already completely broken (even if industry is unable to move on quickly enough) is arguably low, but still the efforts to further improve the theoretical results are probably considerable and it is unclear how much scientific value might be yielded by the engineering efforts to achieve what industry deems a "practical" attack.
This is the classic set of conditions in which white hat teams have very little incentive while black hat teams have all the incentive to keep advancements secret.

---

Anyway, I believe for most users the details of how practical a preimage attack is when collision attacks are practical is of little relevance.
